rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,634 @@
+
+The auth server generating the keys will then use the configured CA keystore to
+generate an `RSA` keypair used to wrap the `Identity`. Once encrypted, the
+`Recipient`, wrapped `Identity`, and wrapping keypair are added as a new entry
+to the `recording_encryption.spec.active_keys` list.
+
+`RSA` key generation is already supported for signing use-cases across AWS KMS,
+GCP CKM, and PKCS#11. However we will need to add support for calling into
+their native decryption functions in order to unwrap `Identity` keys:
+
+- AWS KMS supports setting an algorithm of `RSAES_OAEP_SHA_256` when using the
+  `Decrypt` action of the KMS API.
+- GCP CKM supports generating keys of type `RSAES_OAEP_2048_SHA256` which can
+  be used when calling the `AsymmetricDecrypt` API from their go SDK. Worth
+  noting that keys generated for signing can not be used for decryption, so key
+  generation with a GCP backend will need to be modified slightly.
+- This should be supported by most HSMs through the `CKM_RSA_PKCS_OAEP` and
+  `CKM_SHA256` mechanisms which can be passed to the `C_DecryptInit` function
+  exposed by PKCS#11.
+
+In order to collaboratively generate and share the `Identity` and `Recipient`
+keys, all auth servers must create a watcher for `Put` events against the
+`RecordingEncryption` resource. Modifications to this resource will signal
+existing auth servers to investigate whether or not there is work that needs to
+be done. For example, when adding a new auth server to an environment, it will
+find that there is already an `X25519` keypair configured. It will check if any
+active keys are accessible (detailed in the next paragraph) and, in the case
+that there are none, generate a new `RSA` wrapping keypair. The new keypair
+will be added as an entry to the list of active keys but without a wrapped
+`Identity`. Any other auth server with an active key can inspect the new entry,
+unwrap their own copy of the `Identity`, and wrap it again using software OAEP
+encryption with the entry's public `RSA` key. The re-wrapped `Identity` is then
+saved as the wrapped key for the new entry and the newly added auth server will
+be able to decrypt sessions.
+
+Each time there is a change to the active keys, the set of public keys will
+also be assigned to `session_recording_config.status.encryption_keys` to be
+used by nodes throughout the cluster.
+
+When using KMS keystores, auth servers may share access to the same key. In
+that case, they will also share the same wrapped key which can be identified by
+the key ID. For HSMs integrated over PKCS#11, the auth server's host UUID is
+attached to the key ID and will be used to determine access.
+
+In order to avoid unintended automatic deletion, keys provisioned for encrypted
+session recording will be tagged with a different label where applicable. This
+will prevent older auth servers from deleting this new keytype and allow new
+auth servers to identify which keys are no longer in use before deletion.
```

Comment:
Which keys are we referring to here? The key encryption keys or the recording encryption keys? It's not clear if the KEKs are going to be accumulated forever or if we're going to re-encrypt the REKs as their rotation happens (as well as encrypting new REKs as they get rotated in).